sycl/doc/extensions/Barrier/Barrier.asciidoc

```diff
@@ -0,0 +1,184 @@
+
+`arrival_token arrive_and_drop_no_complete(int32_t count)` : Reduces expected arrival count for future cycles by one, executes arrival operation that counts as `count` arrivals and returns a token that is needed for the wait call corresponding to this arrival. This must not be the last arrival thet causes the cycle to complete - it would be undefined behavior.
+
+`void arrive_copy_async()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete. Before the arrive operation is triggered, the pending count on the barrier is increased by 1, so after the arrival there is no change to the pending count.
+
+`void arrive_copy_async_no_inc()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete.
+
+`void wait(arrival_token arrival)` : Executes wait operation, blocking until the predetermined number of work items reach arrive.
+
+`void test_wait(arrival_token arrival)` : Checks whether all the arrivals have already happened for the current cycle, returning `true` if they did and `false` if `wait(arrival)` would block.
+
+`static constexpr uint64_t max()` : Returns the maximum value of the expected and pending counts supported by the implementation. Exceeding this in the internal state of the barrier is undefined behavior.
+
+==== Sample Header
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+class barrier {
+  [implementation defined internal state]
+
+public:
+  using arrival_token = [implementation defined];
+
+  // barriers cannot be moved or copied
+  barrier(const barrier &other) = delete;
+  barrier(barrier &&other) noexcept = delete;
+  barrier &operator=(const barrier &other) = delete;
+  barrier &operator=(barrier &&other) noexcept = delete;
+
+  void initialize(uint32_t expected_count);
+  void invalidate();
+  arrival_token arrive();
+  arrival_token arrive_and_drop();
+  arrival_token arrive_no_complete(int32_t count);
+  arrival_token arrive_and_drop_no_complete(int32_t count);
+  void arrive_copy_async();
+  void arrive_copy_async_no_inc();
+  void wait(arrival_token arrival);
+  void test_wait(arrival_token arrival);
+  arrival_token arrive_and_wait();
+  static constexpr uint64_t max();
+};
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+== Issues
+
+None.
+
+//. asd
+//+
+//--
+//*RESOLUTION*: Not resolved.
+//--
+
+. Is `barrier` the best name? Reasons for that name are that it is mostly in line with c+\+20 `std::barrier` and CUDA has the same name for this functionality. However it might be confusing with `group_barrier`, which is not present in c++20 and has a different name in CUDA - `__syncthreads`. Earlier version of CUDA docs called this `awbarrier`. Now that name is deprecated and they call it asynchronous barrier in text and `barrier` in code. Related PTX instructions use `mbarrier`.
```

Comment:
`barrier` does seem to make sense here, given that its design is aligned with `std::barrier` and the [cuda::barrier](https://nvidia.github.io/libcudacxx/extended_api/synchronization_primitives/barrier.html) from libcu++.  Other names I've heard for this concept are "non-blocking barriers" and "split barriers".

Reply:
Added other ideas